src/main/java/rlbot/manager/BotManager.java

```diff
@@ -86,22 +88,48 @@ public void ensureStarted() {
     }
 
     private void doLoop() {
-        while (keepRunning) {
+        // Minimum call rate when paused.
+        final long MAX_AGENT_CALL_PERIOD = 1000 / 30;
+        final float TAREHARTS_CONSTANT = 1f;
+
+        long lastCallRealTime = System.currentTimeMillis();
 
+        float lastTickGameTime = -1;
+        float frameUrgency = 0;
+
+        while (keepRunning) {
+            // Python version: https://github.com/RLBot/RLBot/blob/master/src/main/python/rlbot/botmanager/bot_manager.py#L194-L212
+            final int refreshRate = this.refreshRate.get();
             try {
+                // Retrieve latest packet
                 latestPacket = RLBotDll.getFlatbufferPacket();
-                synchronized (dinnerBell) {
-                    dinnerBell.notifyAll();
+
+                // Run the bot only if gameInfo has updated
+                final float tickGameTime = latestPacket.gameInfo().secondsElapsed();
+                final long now = System.currentTimeMillis();
+                final boolean shouldCallWhilePaused = now - lastCallRealTime >= MAX_AGENT_CALL_PERIOD;
+
+                if(frameUrgency < TAREHARTS_CONSTANT / refreshRate) // Urgency increases every frame, but don't let it build up a large backlog
```

Comment:
These lines still seem redundant to me. I believe you could replace these two lines with just `frameUrgency = Math.min(frameUrgency + (tickGameTime - lastTickGameTime), TAREHARTS_CONSTANT / refreshRate);`

I might be wrong though. I'm about to come through and add https://github.com/RLBot/RLBot/commit/3d4ccfeb5f306679ea574e9daa76c0d4521783a0, so I'll experiment with it at that point.